sharktank/sharktank/layers/paged_attention.py

```diff
@@ -144,6 +144,11 @@ def read(
         key = selected[:, :, 0, :].flatten(1, 2)
         value = selected[:, :, 1, :].flatten(1, 2)
 
+        if self.devices:
+            # Explicitly passing a list of one value to avoid redundant transfer inside ReplicateTensor.__init__.
+            key = ReplicatedTensor(ts=[key], devices=self.devices)
+            value = ReplicatedTensor(ts=[value], devices=self.devices)
+
```

Comment:
Why are sharding details exposed in KVCache when there is a ShardedKVCache class?

Reply:
ShardedKVCache is used for tensor parallelism. For pipeline parallelim we have the PipelinedCache which uses both KVCache (PP only) and ShardedKVCache (TP and PP)